### PR TITLE
Optimize elliptic solver (5): Multigrid algorithm

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Two optimizations of the multigrid algorithm.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp [hunk 1]
```diff
@@ -38,8 +39,12 @@ namespace LinearSolver::multigrid::detail {
 /// \cond
 template <typename FieldsTag, typename OptionsGroup, typename SourceTag>
 struct SendCorrectionToFinerGrid;
+template <typename FieldsTag, typename OptionsGroup, typename SourceTag>
+struct SkipPostsmoothingAtBottom;
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp]: `Smoothing` (capitalization)

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp [hunk 2]
```diff
@@ -43,6 +43,33 @@ struct OutputVolumeData {
...
+struct EnablePreSmoothing {
+  static std::string name() { return "PreSmoothing"; }
+  using type = bool;
+  static constexpr Options::String help =
+      "Set to 'False' to disable pre-smoothing altogether (\"cascading "
+      "multigrid\"). Note that pre-smoothing can be necessary to remove "
+      "high-frequency modes in the data that get restricted to coarser grids, "
+      "since such high-frequency modes can introduce aliasing. However, when "
+      "running only a single V-cycle as preconditioner, the initial field is "
+      "typically zero, so pre-smoothing may not worthwile.";
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp]: "may not be"

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp [hunk 3]
```diff
@@ -43,6 +43,33 @@ struct OutputVolumeData {
...
+template <typename OptionsGroup>
+struct EnablePostSmoothingAtBottom {
+  static std::string name() { return "PostSmoothingAtBottom"; }
+  using type = bool;
+  static constexpr Options::String help =
+      "Set to 'False' to skip post-smoothing on the coarsest grid. This means "
+      "only pre-smoothing runs on the coarsest grid, so the coarsest grid "
+      "experiences less smoothing altogether. This is typically only "
+      "desireable if the coarsest grid covers the domain with a single "
+      "element, or very few, so smoothing is exceptionally effective.";
```
Comment (wthrowe) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp]: "This is typically only desirable..." is "this" setting the option or setting the option to false (as discussed in the previous sentences)?  It seems to me to mean "Setting this option to false is typically only desirable", but then it seems weird to disable smoothing if "smoothing is exceptionally effective".

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp (hunk 1 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/LinearSolver/Multigrid/ElementActions.hpp]: Added a commit for this

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp (hunk 3 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp]: expanded the text a bit

Comment (wthrowe) [reply src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp]: That clears it up.  Thanks.

**Review by wthrowe — COMMENTED** [review]
Squash.

**Review by wthrowe — DISMISSED** [review]

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp [hunk 4]
```diff
@@ -43,6 +43,34 @@ struct OutputVolumeData {
...
 };
 
+template <typename OptionsGroup>
+struct EnablePreSmoothing {
+  static std::string name() { return "PreSmoothing"; }
+  using type = bool;
+  static constexpr Options::String help =
+      "Set to 'False' to disable pre-smoothing altogether (\"cascading "
+      "multigrid\"). Note that pre-smoothing can be necessary to remove "
+      "high-frequency modes in the data that get restricted to coarser grids, "
```
Comment (nikwit) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:53]: in the data that get(s) ? I assume the data gets restricted and not the high-frequency modes in this sentence.

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp [hunk 5]
```diff
@@ -43,6 +43,34 @@ struct OutputVolumeData {
...
+
+template <typename OptionsGroup>
+struct EnablePostSmoothingAtBottom {
+  static std::string name() { return "PostSmoothingAtBottom"; }
+  using type = bool;
+  static constexpr Options::String help =
+      "Set to 'False' to skip post-smoothing on the coarsest grid. This means "
+      "only pre-smoothing runs on the coarsest grid, so the coarsest grid "
+      "experiences less smoothing altogether. This is typically only "
+      "desireable if the coarsest grid covers the domain with a single "
```
Comment (nikwit) [inline src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp]: desirable

**Review by nikwit — CHANGES_REQUESTED** [review]
small typos, can squash directly

File: src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp (hunk 4 above)
Comment (nilsvu) [reply src/ParallelAlgorithms/LinearSolver/Multigrid/Tags.hpp:53]: either way I think `get` is correct because `data` is always plural 🤓

**Comment by nilsvu** [comment]
Rebased and squashed in the typo fix. Thanks for reviewing @wthrowe and @nikwit !

**Review by nikwit — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**